### ISSUE TITLE
Conv Kernel Shape

### ISSUE THREAD

**Issue by tjingrant** [onnx/onnx-tensorflow#95]
@fumihwh I'm working on the tutorial but it seems like we have some problem recognizing the filter kernel size in conv2d. 

https://github.com/onnx/onnx-tensorflow/blob/552a87387d93b878bdf79ca25896173a74f14aa4/onnx_tf/frontends/frontend_v1.py#L84

Tensorflow filter size is specified as such:
[filter_height, filter_width, in_channels, out_channels]

Thus seems like we should be taking the first d elements instead of the current approach. Maybe we should change it?

**Comment by fumihwh** [comment]
@tjingrant 
https://github.com/onnx/onnx-tensorflow/blob/552a87387d93b878bdf79ca25896173a74f14aa4/onnx_tf/frontend.py#L189-L192
I transposed kernel first.

**Comment by tjingrant** [comment]
Ah, so I explicitly created the variable and the kernel is not present in the variable name. This is actually quite common if one intends to share the same weight between many layers.

In addition, we should have talked about this during code review for an entirely different reason: if we create a scope named "kernel" in TF then every constant in this scope will be transposed by this piece of code...

Basically I don't think this is the right way to capture convolution kernel... What do you think?

**Comment by tjingrant** [comment]
So I managed to get mnist working (with some ad-hoc fix for the kernel transpose). Turns out bias_add needs more hint about broadcast axis and a fix is already submitted.

The reason why I gave up ResNet is that TF ResNet is trained in NHWC mode. I'll have to think about how to import NCHW based models (it's quite complicated because there is only NCHW conv in ONNX). Please comment if you have any thought on this issue.

**Comment by fumihwh** [comment]
Sorry, I can't get it.
Export TF ResNet to onnx failed because of there is only NCHW conv in ONNX?

**Comment by tjingrant** [comment]
I wouldn't say it failed but to get it working more (non-trivial) work is needed. 
Last time I checked, there is no NHWC conv in ONNX.

**Comment by fumihwh** [comment]
From pr's comment,
I think change the consts directly and put transposed kernel into onnx graph (as initializer) is a better way.

**Comment by tjingrant** [comment]
Yeah but here it's not about the kernel that we need to transpose. It's the input feature map that we need to transpose. It's an issue different from the one the PR tries to address.

**Comment by fumihwh** [comment]
Now I got. Transposing input at first and output at last (also with NCHW conv and other ops) should work?

**Comment by tjingrant** [comment]
Yes but performance implication is huge in that case.

**Comment by fumihwh** [comment]
Is that huge?

**Comment by tjingrant** [comment]
There are two potential fixes along the line we were talking.
1. Transpose input at first and output at last for the entire model.
2. Transpose input at first and output at last for each op.

My thought was the second. I suspect your thought was the first. The reason why I don't think first is a good idea is because the engineering cost is very high and the logic behind complex. Essentially you'll need each op handler to be able to react to a transposition of indices. 

I think for the second option the performance is obviously very bad... But second thought, the ONNX optimizer may help us go from 2->1 automatically.

**Comment by fumihwh** [comment]
2, will ONNX optimizer do that for us?

**Comment by tjingrant** [comment]
@fumihwh 1 puts us in a familiar situation: how do you tell which tensors are input (thus to be transposed), which tensors are not (thus not to be transposed)? After all, a tensor can be both according to the current spec.

**Comment by fumihwh** [comment]
annoying.....current I always train again from scratch using NCHW...
